Enable mixed configuration of skip-[inbound|outbound]-ports

javaducky opened the pull request:
**Subject**
Enable mixed configuration of `skip-[inbound|outbound]-ports` using port numbers and ranges

**Problem**
The ability to configure port ranges for the `skip-[inbound|outbound]-ports` setting was requested.  The existing functionality allowed for only a comma-separated list of port numbers.

**Solution**
A new method has been created which will accept the mixed use of port numbers and ranges for the noted configuration settings.  As an example, `--skip-inbound-ports 8-10,15,17` may be provided via CLI which would allow for the configuration to be applied to ports 8, 9, 10, 15, and 17.  Unit tests have also been provided for this method.

Overall solution depends upon changes to the `linkerd2-proxy-init` with https://github.com/linkerd/linkerd2-proxy-init/pull/7.

**Validation**
Created a new unit tests to confirm desired functionality with applicable success and failure scenarios.

Fixes #3752 

Signed-off-by: Paul Balogh <javaducky@gmail.com>

zaharidichev commented on cli/cmd/root.go:
```diff
@@ -265,6 +268,18 @@ func (options *proxyConfigOptions) validate() error {
 			options.proxyLogLevel)
 	}
 
+	if options.rawIgnoreInboundPorts != nil {
```
I think (correct me if wrong) this function only validates things, but now it not only validates it but it mutates the options as well. So we can either move that logic to another function or simply rename it to `parseAndValidate` ?

javaducky replied:
Not a problem to change the name, I hadn't really thought of it that way but it does make sense.

zaharidichev commented on cli/cmd/root.go:
```diff
@@ -274,6 +289,45 @@ func registryOverride(image, registry string) string {
+//   []string{"23","25-27"}
+// will result in
+//   []uint{23,25,26,27}
+func mapRangeSlice(target *[]uint, source []string) error {
```
I wonder whether it makes sense to extract that into a separate file. Having a `root_test.go` that only tests this functionality seems a bit odd to me. WDYT?

javaducky replied:
I had actually considered a separate file at first, but was worried about starting a utility method "junk drawer."  I do like the idea of creating a `range_slice.go` however which would be similar to how the `pflags` is organized with `string_slice.go` and `uint_slice.go`.  Does `range_slice.go` or `port_range_slice.go` sound better?

zaharidichev replied:
I think `range_slice.go` sounds good. "Utility method junk drawer" - that is a good one !

zaharidichev commented on cli/cmd/root.go:
```diff
@@ -274,6 +289,45 @@ func registryOverride(image, registry string) string {
+				return fmt.Errorf("\"%s\": upper-bound must be greater than lower-bound", portOrRange)
+			}
+			for i := lower; i <= upper; i++ {
+				*target = append(*target, uint(i))
```
+1 For testing the bounds

javaducky replied:
Thanks!  This also made me realize that I did not include a test for the upper- and lower-bounds being equal.

zaharidichev commented on cli/cmd/root.go:
```diff
@@ -284,8 +338,8 @@ func (options *proxyConfigOptions) flagSet(e pflag.ErrorHandling) *pflag.FlagSet
-	flags.UintSliceVar(&options.ignoreInboundPorts, "skip-inbound-ports", options.ignoreInboundPorts, "Ports that should skip the proxy and send directly to the application")
-	flags.UintSliceVar(&options.ignoreOutboundPorts, "skip-outbound-ports", options.ignoreOutboundPorts, "Outbound ports that should skip the proxy")
+	flags.StringSliceVar(&options.rawIgnoreInboundPorts, "skip-inbound-ports", options.rawIgnoreInboundPorts, "Ports and/or port ranges that should skip the proxy and send directly to the application")
+	flags.StringSliceVar(&options.rawIgnoreOutboundPorts, "skip-outbound-ports", options.rawIgnoreOutboundPorts, "Outbound ports and/or port ranges that should skip the proxy")
```
Might be worth to specify that these port ranges are inclusive of lower and upper bound. Off by one errors are the root of the rest of all evil :)

javaducky replied:
Was thinking of using `Ports and/or port ranges (inclusive) ...` or `Ports and/or inclusive port ranges ...`.  Any preference?  My initial lean was toward the former.

javaducky commented:
Also, @zaharidichev thank you so much for the quick review feedback!  I truly appreciate it!

grampelberg commented:
Let's make sure to land the required proxy-init changes before merging this PR.

javaducky commented:
> Let's make sure to land the required proxy-init changes before merging this PR.

Do you have a feature branch you're looking to get this rebased on?  If that would help, I'm more than happy to oblige.

grampelberg commented:
@javaducky I had assumed that https://github.com/linkerd/linkerd2-proxy-init would require changes, but looking at this now it seems like that won't be an issue.

ihcsim commented:
@javaducky Thanks again for working on this. I wonder if there is a better way to handle cases wotj wide port ranges. E.g., if I run this command on your branch
```
bin/go-run cli install --skip-inbound-ports=1000-5000 | less
```

My `linkerd-config` config map YAML looks like:
```
proxy: |
    {"proxyImage":{"imageName":"gcr.io/linkerd-io/proxy","pullPolicy":"IfNotPresent"},"proxyInitImage":{"imageName":"gcr.io/linkerd-io/proxy-init","pullPolicy":"IfNotPresent"},"controlPort":{"port":4190},"ignoreInboundPorts":[{"port":1000},{"port":1001},{"port":1002},{"port":1003},{"port":1004},{"port":1005},{"port":1006},{"port":1007},{"port":1008},{"port":1009},{"port":1010},{"port":1011},{"port":1012},{"port":1013},{"port":1014},{"port":1015},{"port":1016},{"port":1017},{"port":1018},{"port":1019},{"port":1020},{"port":1021},{"port":1022},{"port":1023},{"port":1024},{"port":1025},{"port":1026},{"port":1027},{"port":1028},{"port":1029},{"port":1030},{"port":1031},{"port":1032},{"port":1033},{"port":1034},{"port":1035},{"port":1036},{"port":1037},{"port":1038},{"port":1039},{"port":1040},{"port":1041},{"port":1042},{"port":1043},{"port":1044},{"port":1045},{"port":1046},{"port":1047},{"port":1048},{"port":1049},{"port":1050},{"port":1051},{"port":1052},{"port":1053},{"port":1054},{"port":1055},{"port":1056},{"port":1057},{"port":1058},{"port":1059},{"port":1060},{"port":1061},{"port":1062},{"port":1063},{"port":1064},{"port":1065},{"port":1066},{"port":1067},{"port":1068},{"port":1069},{"port":1070},{"port":1071},{"port":1072},{"port":1073},{"port":1074},{"port":1075},{"port":1076},{"port":1077},{"port":1078},{"port":1079},{"port":1080},{"port":1081},{"port":1082},{"port":1083},{"port":1084},{"port":1085},{"port":1086},{"port":1087},{"port":1088},{"port":1089},{"port":1090},{"port":1091},{"port":1092},{"port":1093},{"port":1094},{"port":1095},{"port":1096},{"port":1097},{"port":1098},{"port":1099},{"port":1100},{"port":1101},{"port":1102},{"port":1103},{"port":1104},{"port":1105},{"port":1106},{"port":1107},{"port":1108},{"port":1109},{"port":1110},{"port":1111},{"port":1112},{"port":1113},{"port":1114},{"port":1115},{"port":1116},{"port":1117},{"port":1118},{"port":1119},{"port":1120},{"port":1121},{"port":1122},{"port":1123},{"port":1124},{"port":1125},{"port":1126},{"port":1127},{"port":1128},{"port":1129},{"port":1130},{"port":1131},{"port":1132},{"port":1133},{"port":1134},{"port":1135},{"port":1136},{"port":1137},{"port":1138},{"port":1139},{"port":1140},{"port":1141},{"port":1142},{"port":1143},{"port":1144},{"port":1145},{"port":1146},{"port":1147},{"port":1148},{"port":1149},{"port":1150},{"port":1151},{"port":1152},{"port":1153},{"port":1154},{"port":1155},{"port":1156},{"port":1157},{"port":1158},{"port":1159},{"port":1160},{"port":1161},{"port":1162},{"port":1163},{"port":1164},{"port":1165},{"port":1166},{"port":1167},{"port":1168},{"port":1169},{"port":1170},{"port":1171},{"port":1172},{"port":1173},{"port":1174},{"port":1175},{"port":1176},{"port":1177},{....
```

The same goes for my `proxy-init` container YAML:
```
initContainers:
      - args:
        - --incoming-proxy-port
        - "4143"
        - --outgoing-proxy-port
        - "4140"
        - --proxy-uid
        - "2102"
        - --inbound-ports-to-ignore
        - 4190,4191,1000,1001,1002,1003,1004,1005,1006,1007,1008,1009,1010,1011,1012,1013,1014,1015,1016,1017,1018,1019,1020,1021,1022,1023,1024,1025,1026,1027,1028,1029,1030,1031,1032,1033,1034,1035,1036,1037,1038,1039,1040,1041,1042,1043,1044,1045,1046,1047,1048,1049,1050,1051,1052,1053,1054,1055,1056,1057,1058,1059,1060,1061,1062,1063,1064,1065,1066,1067,1068,1069,1070,1071,1072,1073,1074,1075,1076,1077,1078,1079,1080,1081,1082,1083,1084,1085,1086,1087,1088,1089,1090,1091,1092,1093,1094,1095,1096,1097,1098,1099,1100,1101,1102,1103,1104,1105,1106,1107,1108,1109,1110,1111,1112,1113,1114,1115,1116,1117,1118,1119,1120,1121,1122,1123,1124,1125,1126,1127,1128,1129,1130,1131,1132,1133,1134,1135,1136,1137,1138,1139,1140,1141,1142,1143,1144,1145,1146,1147,1148,1149,1150,1151,1152,1153,1154,1155,1156,1157,1158,1159,1160,1161,1162,1163,1164,1165,1166,1167,1168,1169,1170,1171,1172,1173,1174,1175,1176,1177,1178,1179,1180,1181,1182,1183,1184,1185,1186,1187,1188,1189,1190,1191,1192,1193,1194,1195,1196,1197,1198,1199,1200,1201,1202,1203,1204,1205,1206,1207,1208,1209,1210,1211,1212,1213,1214,1215,1216,1217,1218,1219,1220,1221,1222,1223,1224,1225,1226,1227,1228,1229,1230,1231,1232,1233,1234,1235,1236,1237,1238,1239,1240,1241,1242,1243,1244,1245,1246,1247,1248,1249,1250,1251,1252,1253,1254,1255,1256,1257,1258,1259,1260,1261,1262,1263,1264,1265,1266,1267,1268,1269,1270,1271,1272,1273,1274,1275,1276,1277,1278,1279,1280,1281,1282,1283,1284,1285,1286,1287,1288,1289,1290,1291,1292,1293,1294,1295,1296,1297,1298,1299,1300,1301,1302,1303,1304,1305,1306,1307,1308,1309,1310,1311,1312,1313,1314,1315,1316,1317,1318,1319,1320,1321,1322,1323,1324,1325,1326,1327,1328,1329,1330,1331,1332,1333,1334,1335,1336,1337,1338,1339,1340,1341,1342,1343,1344,1345,1346,1347,1348,1349,1350,1351,1352,1353,1354,1355,1356,1357,1358,1359,1360,1361,1362,1363,1364,1365,1366,1367,1368,1369,1370,1371,1372,1373,1374,1375,1376,1377,1378,1379,1380,1381,1382,1383,1384,1385,1386,1387,1388,1389,1390,1391,1392,1393,1394,1395,1396,1397,1398,1399,1400,1401,1402,1403,1404,1405,1406,1407,1408,1409,1410,1411,1412,1413,1414,1415,1416,1417,1418,1419,1420,1421,1422,1423,1424,1425,1426,1427,1428,1429,1430,1431,1432,1433,1434,1435,1436,1437,1438,1439,1440,1441,1442,1443,1444,1445,1446,1447,1448,1449,1450,1451,1452,1453,1454,1455,1456,1457,1458,1459,1460,1461,1462....
```

I think this will clutter the outputs of commands like `kubectl [describe|get]`. Is there a way to expand the range without encoding it in the YAML? Or will letting the proxy expand the range be a viable option (which will require some helps from the proxy folks)?

javaducky commented:
@ihcsim Yeah, a wide range definitely makes the YAML uglier than normal.  I think if the underlying configuration were kept as a string slice in the CLI, the range could then be passed along to the proxy easily for expansion.  I'll look further into that and the Rust code to see what I can do.  Honestly, this feels like the better thing to do anyway.

ihcsim commented:
> it would appear the expansion of ranges would be done within the linkerd2-proxy-init repo as this would affect the iptables.

ha... now I see what @grampelberg was talking about earlier. Wanna give that a try?

ihcsim commented:
The other thing we need to take into account is also updating the `config.linkerd.io/skip-[inbound|outbound]-ports` annotations to do the same. LMK if we want to tackle that in this PR or separately. The relevant code is [here](https://github.com/linkerd/linkerd2/blob/a564609ed573c341211092e46a837d4680983b77/pkg/inject/inject.go#L840-L864), which essentially is also what the CLI calls.

javaducky commented:
Not a problem.  I can take care of the proxy-init as well as the annotations.

javaducky commented:
@ihcsim @alpeb As @grampelberg eluded to in a previous comment, there are `linkerd2-proxy-init` changes that *must be merged prior* to this PR.  The PR for the `proxy-init` is https://github.com/linkerd/linkerd2-proxy-init/pull/7 .

As a summary of the most recent changes, we'll no longer hold the configuration for the ignored ports as `uint`s; instead, we'll hold onto them as `string`s, however the values will be validated to ensure only good values are accepted.  Those string representations will be passed into the `proxy-init` image as arguments where they will actually be utilized as `int`s when configuring the `iptables`.

On another note...the `proxy-init` treats these ports as `int` compared to what the CLI had been treating as `uint`.  I left the use in the `proxy-init` as `int`; validation will ensure there are no negative-values allowed.

javaducky commented:
@zaharidichev given the amount of changes, you'll probably want to revisit this PR :)

zaharidichev commented on cli/cmd/install_test.go:
```diff
@@ -172,6 +181,7 @@ func TestRender(t *testing.T) {
 		{haWithOverridesValues, "install_ha_with_overrides_output.golden"},
 		{noInitContainerValues, "install_no_init_container.golden"},
+		{withProxyIgnoresValues, "install_proxy_ignores.golden"},
 	}
```
Maybe its worth adding a test that fails and outputs the reason, just to ensure the validation logic you wrote will remain wired correctly into the install cmd

javaducky replied:
Added the new `TestValidateAndBuild_Errors` method to test fail cases.

zaharidichev commented on cli/cmd/range_slice_test.go:
```diff
@@ -0,0 +1,39 @@
+	assertError(t, validateRangeSlice([]string{"not-number"}), "not a valid lower-bound")
+	assertError(t, validateRangeSlice([]string{"-23-25"}), "ranges expected as")
+	assertError(t, validateRangeSlice([]string{"-23"}), "not a valid lower-bound")
+	assertError(t, validateRangeSlice([]string{"25-23"}), "upper-bound must be greater than or equal to")
```
Really like this extensive testing

javaducky replied:
_Hopefully_ I improved the usability of the tests.  Previously, any failure really didn't provide the context as to which case failed.  Updated for parameterized testing which will indicate the specific test case that caused the failure.

alpeb commented on cli/cmd/main_test.go:
```diff
@@ -14,7 +14,7 @@ var (
 	updateFixtures bool
 
-	// prettyDiff is set by the `-verbose-diff` flag.
+	// prettyDiff is set by the `-pretty-diff` flag.
```
Thanks :+1:

javaducky commented:
Ok guys...I think I've got it to what we're looking for.  Hopefully you squash commits when merging 😁 .

These last couple of commits, I thought I was gettin fancier with the parameterized tests, but _scopelinter_ was having none of it.  Not sure if the same build rules should be ported over to the `linkerd-proxy-init` as it has several such uses for parameterized tests.

I've rebased everything to the latest `master` branch, so things should go smooth.

Please let me know if there's anything else!  I'll be sifting through the issues list.

zaharidichev commented on cli/cmd/install_test.go:
```diff
@@ -173,6 +182,7 @@ func TestRender(t *testing.T) {
 		{haValues, "install_ha_output.golden"},
 		{haWithOverridesValues, "install_ha_with_overrides_output.golden"},
 		{noInitContainerValues, "install_no_init_container.golden"},
+		{withProxyIgnoresValues, "install_proxy_ignores.golden"},
```
Aweseome! Thanks !

alpeb commented:
@javaducky as a last change, could you please update the `ProxyInitVersion` value in `/pkg/version/version.go` to `v1.3.0` and the corresponding test fixtures changes?
Note the CI tests will fail till we publish that version; I'll keep an eye on this one to re-run the CI once that happens.

alpeb commented:
btw, I tested the port-skipping was indeed working using emojivoto, with a `v1.3.0` proxy-init that I published locally from linkerd/linkerd2-proxy-init#7
```bash
$ bin/go-run cli install --init-image-version v1.3.0 | k apply -f -
$ k apply -f https://run.linkerd.io/emojivoto.yml
$ k -n emojivoto get deployments web -oyaml|bin/go-run cli inject -|k apply -f -
$ bin/go-run cli dashboard
```
Which results in `web` properly showing stats:
![image](https://user-images.githubusercontent.com/554287/70923557-08e71800-1ff6-11ea-84af-819d5449a53d.png)

And then injecting again ignoring its inbound port using a range:
```bash
k -n emojivoto get deployments web -oyaml|bin/go-run cli inject --skip-inbound-ports 8079-8081 -|k apply -f -
```
![image](https://user-images.githubusercontent.com/554287/70923778-709d6300-1ff6-11ea-853e-04f0da6d39fd.png)
So we no longer see stats but traffic is clearly going through as shown in the downstream dependencies :+1:

javaducky commented:
Should be all ready for the release of `v1.3.0` of `proxy-init`.

javaducky commented:
@alpeb Looks like `v1.3.0` has been cut and image is now available in registry.  Should be good to trigger build again.

alpeb commented:
@javaducky given `go.mod` was changed in the last commit, could you also please run `bin/update-go-deps-shas` to update the Dockerfiles? After that, the CI should go green :heavy_check_mark:

javaducky commented:
@alpeb sorry about the commit mess.  In summary, I've rebased on latest master, fixed version to `1.3.0` for the many files I had missed the first time, updated the SHAs on the `go.sum`, and fixed some post-rebase issues.  Build is all green again 👍

alpeb commented:
Merged! Thanks again @javaducky for the quick turnaround, thorough unit testing, and overall awesomeness! :medal_military: :1st_place_medal:

javaducky commented:
My pleasure...looking forward to picking up on more issues!
